File: Frontend-BAP/screens/VerifyProductsScreen.js
import React, { useState,useEffect } from 'react';
import { Platform ,View, Text, Image, ScrollView, TouchableOpacity } from 'react-native';
import tw from 'tailwind-react-native-classnames';
import { SERVER_URL,mobile_url } from '@env';

const VerifyProductsScreen = ({ route, navigation }) => {
  const { bpp_response } = route.params;
   console.log(bpp_response);
  if (!bpp_response?.message?.order) {
    return (
      <View style={tw`flex-1 justify-center items-center`}>
        <Text style={tw`text-red-500 text-lg`}>❌ Invalid order data</Text>
      </View>
    );
  }
  useEffect(() => {
  callBapLogistics();
}, []);
  
  const [selectedDelivery, setSelectedDelivery] = useState(null);
  const [logisticsOptions, setLogisticsOptions] = useState([]);
  const order = bpp_response.message.order;
  // console.log(order);
  const items = order.items || [];
  const provider = order.provider || {};
  const quote = order.quote || {};
  const fulfillment = order.fulfillment || {};
  const payment = order.payment || {};

  const calculateTotalWeight = () => {
  return items.reduce((total, item) => {
    const perUnitWeight = parseFloat(item.quantity?.unitized?.measure?.value || 0); // weight per unit
    const qty = parseInt(item.quantity?.count || 0);
    return total + (perUnitWeight * qty);
  }, 0);
};


  const callBapLogistics = async () => {
  try {
    // const totalWeight = calculateTotalWeight();
    const payload = {
      start: {
        gps: fulfillment.start?.location?.gps || "13.0827,80.2707",
        address: fulfillment.start?.location?.address || "Warehouse, Chennai"
      },
      end: {
        gps: fulfillment.end?.location?.gps || "12.9716,77.5946",
        address: fulfillment.end?.location?.address || "Customer Address, Bangalore"
      },
      weight_kg: 5
    };

    const response = await fetch(`${SERVER_URL}/logistic/search`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload)
    });

    const data = await response.json();
    console.log("✅ Logistics API Response:", data);

    const providers = data?.message?.catalog?.providers || [];
    setLogisticsOptions(providers); // store in state

  } catch (error) {
    console.error("❌ Logistics API Error:", error);
  }
};

  

  return (
  <View style={tw`flex-1 bg-green-50 relative`}>
    <ScrollView
      style={tw`px-4 pt-5`}
      contentContainerStyle={{ paddingBottom: 160 }}
      showsVerticalScrollIndicator={false}
    >
      {/* 🌾 Provider Info */}
      <View style={tw`mb-6 bg-green-100 p-5 rounded-2xl shadow-md border border-green-200`}>
        <Text style={tw`text-2xl font-bold text-green-900`}>
          {provider.descriptor?.name || "Unknown Provider"}
        </Text>
        <Text style={tw`text-sm text-green-800 mt-1`} numberOfLines={2}>
          📍 {fulfillment.start?.location?.address || "No address"}
        </Text>
      </View>

      {/* 🚚 Fulfillment Info */}
      <View style={tw`bg-white p-5 rounded-2xl shadow-sm mb-6 border border-green-100`}>
        <Text style={tw`text-lg font-semibold text-green-700 mb-3`}>
          🚚 Fulfillment Info
        </Text>
        <Text style={tw`text-sm text-gray-700 mb-1`}>Type: {fulfillment.type || 'N/A'}</Text>
        <Text style={tw`text-sm text-gray-700 mb-1`}>📍 Pickup: {fulfillment.start?.location?.address || 'N/A'}</Text>
        <Text style={tw`text-sm text-gray-700 mb-1`}>🏁 Delivery: {fulfillment.end?.location?.address || 'N/A'}</Text>
        <Text style={tw`text-sm text-gray-700`}>
          ⏳ ETA: {fulfillment.estimated_delivery ? new Date(fulfillment.estimated_delivery).toLocaleString() : "N/A"}
        </Text>
      </View>

      {/* 🧾 Verified Products */}
      <Text style={tw`text-xl font-bold text-green-900 mb-4`}>
        🧾 Verified Products
      </Text>
      {items.map((item, idx) => {
        const subtotal = (parseFloat(item.price?.value || 0) * (item.quantity?.count || 1)).toFixed(2);
        const unit = item.quantity?.unitized?.measure?.unit || '';
        const isOrganic = item.tags?.[0]?.value === 'true';
        const imageUrl = item.descriptor?.images?.[0] || "https://via.placeholder.com/60";

        return (
          <View key={idx} style={tw`bg-white p-4 rounded-2xl shadow-sm mb-4 border border-green-100`}>
            <View style={tw`flex-row`}>
              {/* 🖼 Image + Organic Badge */}
              <View style={tw`relative mr-4`}>
                <Image
                  source={{ uri: imageUrl }}
                  style={tw`w-16 h-16 rounded-lg`}
                  resizeMode="cover"
                />
                {isOrganic && (
                  <View style={tw`absolute -top-2 -right-2 bg-green-200 px-2 py-0.5 rounded-full shadow`}>
                    <Text style={tw`text-green-900 text-xs font-semibold`}>Organic</Text>
                  </View>
                )}
              </View>

              {/* 🛒 Product Info */}
              <View style={tw`flex-1`}>
                <Text style={tw`font-semibold text-gray-900 text-sm mb-1`}>
                  {item.descriptor?.name}
                </Text>
                <Text style={tw`text-xs text-gray-700 mb-1`}>
                  Qty: {item.quantity?.count} {unit}
                </Text>
                <Text style={tw`text-xs text-gray-700 mb-1`}>
                  Unit Price: ₹{parseFloat(item.price?.value || 0).toFixed(2)}
                </Text>
                <Text style={tw`text-sm text-green-800 font-semibold mt-1`}>
                  Subtotal: ₹{subtotal}
                </Text>
              </View>
            </View>
          </View>
        );
      })}

      {/* 💰 Price Summary */}
      <View style={tw`bg-white p-5 rounded-2xl shadow-md mb-6 border border-green-100`}>
        <Text style={tw`text-lg font-bold text-green-900 mb-3`}>
          💰 Price Summary
        </Text>
        {(quote.breakup || []).map((entry, idx) => (
          <View key={idx} style={tw`flex-row justify-between mb-1`}>
            <Text style={tw`text-sm text-gray-700`}>{entry.title}</Text>
            <Text style={tw`text-sm text-gray-900`}>
              ₹{parseFloat(entry.price?.value || 0).toFixed(2)}
            </Text>
          </View>
        ))}
        <View style={tw`border-t border-green-200 mt-3 pt-3 flex-row justify-between`}>
          <Text style={tw`text-base font-bold text-gray-800`}>Total</Text>
          <Text style={tw`text-base font-bold text-green-900`}>
            ₹{parseFloat(quote.price?.value || 0).toFixed(2)}
          </Text>
        </View>
      </View>

      {/* 🚚 Delivery Options */}
      {logisticsOptions.length > 0 && (
        <View style={tw`bg-white p-5 rounded-2xl shadow-md mb-6 border border-green-100`}>
          <Text style={tw`text-xl font-bold text-green-900 mb-4`}>
            Delivery Options
          </Text>

          {logisticsOptions.map((provider, idx) => (
            <View key={idx} style={tw`mb-5`}>
              <View style={tw`bg-white rounded-xl shadow-sm border border-green-200 p-4`}>
                <Text style={tw`text-lg font-semibold text-gray-800 mb-2`}>
                  {provider.descriptor?.name}
                </Text>

                {(provider.items || []).map((option, oIdx) => {
                  const isSelected =
                    selectedDelivery?.providerId === provider.id &&
                    selectedDelivery?.itemId === option.id;

                  return (
                    <TouchableOpacity
                      key={oIdx}
                      activeOpacity={0.85}
                      onPress={() =>
                        setSelectedDelivery({
                          providerId: provider.id,
                          providerName: provider.descriptor?.name,
                          itemId: option.id,
                          itemName: option.descriptor?.name,
                          price: option.price?.value,
                          eta: option.time?.duration
                        })
                      }
                      style={tw`mt-3 rounded-xl p-4 shadow-sm border ${
                        isSelected
                          ? "bg-green-100 border-green-500"
                          : "bg-green-50 border-green-200"
                      }`}
                    >
                      <Text style={tw`text-base font-medium text-gray-900`}>
                        {option.descriptor?.name}
                      </Text>

                      <View style={tw`flex-row justify-between mt-2`}>
                        <Text style={tw`text-sm text-gray-700`}>
                          Price: ₹{parseFloat(option.price?.value || 0).toFixed(2)}
                        </Text>
                        <Text style={tw`text-sm text-gray-700`}>
                          ETA: {option.time?.duration?.replace("PT", "").toLowerCase()}
                        </Text>
                      </View>
                    </TouchableOpacity>
                  );
                })}
              </View>
            </View>
          ))}
        </View>
      )}

    </ScrollView>

    {/* ✅ Fixed Bottom CTA */}
    <View
      style={[
        tw`absolute left-0 right-0 bg-green-100 border-t border-green-200 px-4 pt-3`,
        { bottom: Platform.OS === 'ios' ? 28 : 0, paddingBottom: 16, shadowColor: '#000', shadowOpacity: 0.05, shadowRadius: 8, elevation: 8 }
      ]}
    >
      <TouchableOpacity
        style={tw`bg-green-700 py-4 rounded-full shadow-lg`}
        onPress={() => {
          if (selectedDelivery !== null) {
            navigation.navigate('PaymentScreen', { order });
          } else {
            console.log('Please select a delivery option before proceeding.');
          }
        }}
        activeOpacity={0.8}
      >
        <Text style={tw`text-white text-center font-bold text-base`}>
          ✅ Proceed to Payment
        </Text>
      </TouchableOpacity>
    </View>
  </View>
);



};

export default VerifyProductsScreen;

File: demo/lib/screens/verify_products_screen.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;

class VerifyProductsScreen extends StatefulWidget {
  final Map<String, dynamic> bppResponse;
  const VerifyProductsScreen({Key? key, required this.bppResponse})
    : super(key: key);

  @override
  State<VerifyProductsScreen> createState() => _VerifyProductsScreenState();
}

class _VerifyProductsScreenState extends State<VerifyProductsScreen> {
  Map<String, dynamic>? selectedDelivery;
  List<dynamic> logisticsOptions = [];

  @override
  void initState() {
    super.initState();
    callBapLogistics();
  }

  Map<String, dynamic> get order => widget.bppResponse['message']['order'];
  List<dynamic> get items => order['items'] ?? [];
  Map<String, dynamic> get provider => order['provider'] ?? {};
  Map<String, dynamic> get quote => order['quote'] ?? {};
  Map<String, dynamic> get fulfillment => order['fulfillment'] ?? {};

  Future<void> callBapLogistics() async {
    try {
      final payload = {
        'start': {
          'gps': fulfillment['start']?['location']?['gps'] ?? "13.0827,80.2707",
          'address':
              fulfillment['start']?['location']?['address'] ??
              "Warehouse, Chennai",
        },
        'end': {
          'gps': fulfillment['end']?['location']?['gps'] ?? "12.9716,77.5946",
          'address':
              fulfillment['end']?['location']?['address'] ??
              "Customer Address, Bangalore",
        },
        'weight_kg': 5,
      };

      final response = await http.post(
        Uri.parse('YOUR_SERVER_URL/logistic/search'),
        headers: {"Content-Type": "application/json"},
        body: jsonEncode(payload),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        setState(() {
          logisticsOptions = data['message']?['catalog']?['providers'] ?? [];
        });
      } else {
        print('Logistics API error: ${response.body}');
      }
    } catch (e) {
      print('❌ Logistics API Error: $e');
    }
  }

  @override
  Widget build(BuildContext context) {
    if (order.isEmpty) {
      return Scaffold(
        body: Center(
          child: Text(
            '❌ Invalid order data',
            style: TextStyle(color: Colors.red, fontSize: 18),
          ),
        ),
      );
    }

    return Scaffold(
      backgroundColor: Colors.green[50],
      appBar: AppBar(
        title: Text(provider['descriptor']?['name'] ?? 'Unknown Provider'),
        backgroundColor: Colors.green,
      ),
      body: Stack(
        children: [
          Padding(
            padding: const EdgeInsets.only(bottom: 150),
            child: SingleChildScrollView(
              padding: const EdgeInsets.all(16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  // Provider Info
                  Container(
                    padding: const EdgeInsets.all(16),
                    margin: const EdgeInsets.only(bottom: 16),
                    decoration: BoxDecoration(
                      color: Colors.green[100],
                      borderRadius: BorderRadius.circular(20),
                      border: Border.all(color: Colors.green.shade200),
                    ),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          provider['descriptor']?['name'] ?? 'Unknown Provider',
                          style: const TextStyle(
                            fontSize: 22,
                            fontWeight: FontWeight.bold,
                            color: Colors.green,
                          ),
                        ),
                        const SizedBox(height: 4),
                        Text(
                          "📍 ${fulfillment['start']?['location']?['address'] ?? 'No address'}",
                          style: TextStyle(
                            fontSize: 14,
                            color: Colors.green[800],
                          ),
                        ),
                      ],
                    ),
                  ),

                  // Fulfillment Info
                  Container(
                    padding: const EdgeInsets.all(16),
                    margin: const EdgeInsets.only(bottom: 16),
                    decoration: BoxDecoration(
                      color: Colors.white,
                      borderRadius: BorderRadius.circular(20),
                      border: Border.all(color: Colors.green.shade100),
                    ),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text(
                          '🚚 Fulfillment Info',
                          style: TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                            color: Colors.green[700],
                          ),
                        ),
                        const SizedBox(height: 8),
                        Text('Type: ${fulfillment['type'] ?? 'N/A'}'),
                        Text(
                          '📍 Pickup: ${fulfillment['start']?['location']?['address'] ?? 'N/A'}',
                        ),
                        Text(
                          '🏁 Delivery: ${fulfillment['end']?['location']?['address'] ?? 'N/A'}',
                        ),
                        Text(
                          '⏳ ETA: ${fulfillment['estimated_delivery'] != null ? fulfillment['estimated_delivery'] : 'N/A'}',
                        ),
                      ],
                    ),
                  ),

                  // Verified Products
                  const Text(
                    '🧾 Verified Products',
                    style: TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
                  ),
                  const SizedBox(height: 8),
                  ...items.map((item) {
                    final qty = item['quantity']?['count'] ?? 1;
                    final unit =
                        item['quantity']?['unitized']?['measure']?['unit'] ??
                        '';
                    final unitPrice = double.parse(
                      item['price']?['value']?.toString() ?? '0',
                    );
                    final subtotal = (unitPrice * qty).toStringAsFixed(2);
                    final isOrganic = item['tags']?[0]?['value'] == 'true';
                    final imageUrl =
                        item['descriptor']?['images']?[0] ??
                        'https://via.placeholder.com/60';

                    return Container(
                      padding: const EdgeInsets.all(12),
                      margin: const EdgeInsets.only(bottom: 12),
                      decoration: BoxDecoration(
                        color: Colors.white,
                        borderRadius: BorderRadius.circular(16),
                        border: Border.all(color: Colors.green.shade100),
                      ),
                      child: Row(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Stack(
                            children: [
                              ClipRRect(
                                borderRadius: BorderRadius.circular(12),
                                child: Image.network(
                                  imageUrl,
                                  width: 60,
                                  height: 60,
                                  fit: BoxFit.cover,
                                ),
                              ),
                              if (isOrganic)
                                Positioned(
                                  top: -2,
                                  right: -2,
                                  child: Container(
                                    padding: const EdgeInsets.symmetric(
                                      horizontal: 6,
                                      vertical: 2,
                                    ),
                                    decoration: BoxDecoration(
                                      color: Colors.green[200],
                                      borderRadius: BorderRadius.circular(12),
                                    ),
                                    child: const Text(
                                      'Organic',
                                      style: TextStyle(
                                        fontSize: 10,
                                        fontWeight: FontWeight.bold,
                                        color: Colors.green,
                                      ),
                                    ),
                                  ),
                                ),
                            ],
                          ),
                          const SizedBox(width: 12),
                          Expanded(
                            child: Column(
                              crossAxisAlignment: CrossAxisAlignment.start,
                              children: [
                                Text(
                                  item['descriptor']?['name'] ?? '',
                                  style: const TextStyle(
                                    fontSize: 14,
                                    fontWeight: FontWeight.bold,
                                  ),
                                ),
                                Text(
                                  'Qty: $qty $unit',
                                  style: const TextStyle(fontSize: 12),
                                ),
                                Text(
                                  'Unit Price: ₹$unitPrice',
                                  style: const TextStyle(fontSize: 12),
                                ),
                                Text(
                                  'Subtotal: ₹$subtotal',
                                  style: const TextStyle(
                                    fontSize: 14,
                                    fontWeight: FontWeight.bold,
                                    color: Colors.green,
                                  ),
                                ),
                              ],
                            ),
                          ),
                        ],
                      ),
                    );
                  }).toList(),

                  // Price Summary
                  Container(
                    padding: const EdgeInsets.all(16),
                    margin: const EdgeInsets.only(bottom: 16),
                    decoration: BoxDecoration(
                      color: Colors.white,
                      borderRadius: BorderRadius.circular(16),
                      border: Border.all(color: Colors.green.shade100),
                    ),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        const Text(
                          '💰 Price Summary',
                          style: TextStyle(
                            fontSize: 18,
                            fontWeight: FontWeight.bold,
                            color: Colors.green,
                          ),
                        ),
                        const SizedBox(height: 8),
                        ...((quote['breakup'] ?? []) as List<dynamic>).map((
                          entry,
                        ) {
                          final price = double.parse(
                            entry['price']?['value']?.toString() ?? '0',
                          );
                          return Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: [
                              Text(entry['title'] ?? ''),
                              Text('₹$price'),
                            ],
                          );
                        }),
                        const SizedBox(height: 8),
                        Row(
                          mainAxisAlignment: MainAxisAlignment.spaceBetween,
                          children: [
                            const Text(
                              'Total',
                              style: TextStyle(fontWeight: FontWeight.bold),
                            ),
                            Text(
                              '₹${double.parse(quote['price']?['value']?.toString() ?? '0')}',
                              style: const TextStyle(
                                fontWeight: FontWeight.bold,
                                color: Colors.green,
                              ),
                            ),
                          ],
                        ),
                      ],
                    ),
                  ),

                  // Logistics Options
                  if (logisticsOptions.isNotEmpty) ...[
                    const Text(
                      '🚚 Delivery Options',
                      style: TextStyle(
                        fontSize: 18,
                        fontWeight: FontWeight.bold,
                        color: Colors.green,
                      ),
                    ),
                    const SizedBox(height: 8),
                    ...logisticsOptions.map((provider) {
                      final providerId = provider['id'];
                      return Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          ...((provider['items'] ?? []) as List<dynamic>).map((
                            option,
                          ) {
                            final isSelected =
                                selectedDelivery != null &&
                                selectedDelivery!['providerId'] == providerId &&
                                selectedDelivery!['itemId'] == option['id'];
                            return GestureDetector(
                              onTap: () {
                                setState(() {
                                  selectedDelivery = {
                                    'providerId': providerId,
                                    'providerName':
                                        provider['descriptor']?['name'],
                                    'itemId': option['id'],
                                    'itemName': option['descriptor']?['name'],
                                    'price': option['price']?['value'],
                                    'eta': option['time']?['duration'],
                                  };
                                });
                              },
                              child: Container(
                                margin: const EdgeInsets.only(bottom: 8),
                                padding: const EdgeInsets.all(12),
                                decoration: BoxDecoration(
                                  color: isSelected
                                      ? Colors.green[100]
                                      : Colors.green[50],
                                  borderRadius: BorderRadius.circular(12),
                                  border: Border.all(
                                    color: isSelected
                                        ? Colors.green
                                        : Colors.green.shade200,
                                  ),
                                ),
                                child: Column(
                                  crossAxisAlignment: CrossAxisAlignment.start,
                                  children: [
                                    Text(
                                      option['descriptor']?['name'] ?? '',
                                      style: const TextStyle(
                                        fontWeight: FontWeight.bold,
                                        fontSize: 14,
                                      ),
                                    ),
                                    const SizedBox(height: 4),
                                    Row(
                                      mainAxisAlignment:
                                          MainAxisAlignment.spaceBetween,
                                      children: [
                                        Text(
                                          'Price: ₹${option['price']?['value'] ?? 0}',
                                        ),
                                        Text(
                                          'ETA: ${option['time']?['duration'] ?? ''}',
                                        ),
                                      ],
                                    ),
                                  ],
                                ),
                              ),
                            );
                          }).toList(),
                        ],
                      );
                    }).toList(),
                  ],
                ],
              ),
            ),
          ),

          // Fixed Bottom CTA
          Positioned(
            bottom: 0,
            left: 0,
            right: 0,
            child: Container(
              padding: const EdgeInsets.all(16),
              decoration: BoxDecoration(
                color: Colors.green[100],
                border: Border(top: BorderSide(color: Colors.green.shade200)),
                boxShadow: [
                  BoxShadow(
                    color: Colors.black.withOpacity(0.05),
                    blurRadius: 8,
                  ),
                ],
              ),
              child: ElevatedButton(
                style: ElevatedButton.styleFrom(
                  backgroundColor: Colors.green[700],
                  padding: const EdgeInsets.symmetric(vertical: 16),
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(12),
                  ),
                ),
                onPressed: selectedDelivery != null
                    ? () {
                        Navigator.pushNamed(
                          context,
                          '/paymentScreen',
                          arguments: {'order': order},
                        );
                      }
                    : () {
                        ScaffoldMessenger.of(context).showSnackBar(
                          const SnackBar(
                            content: Text(
                              'Please select a delivery option before proceeding.',
                            ),
                          ),
                        );
                      },
                child: const Text(
                  '✅ Proceed to Payment',
                  style: TextStyle(
                    color: Colors.white,
                    fontWeight: FontWeight.bold,
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
